File: lua/problem_27.lua
local function isBalanced(s)
    local match = { [')'] = '(', [']'] = '[', ['}'] = '{' }
    local stack = {}
    for i = 1, #s do
        local c = s:sub(i, i)
        if c == '(' or c == '[' or c == '{' then
            stack[#stack + 1] = c
        elseif c == ')' or c == ']' or c == '}' then
            if #stack == 0 or stack[#stack] ~= match[c] then return false end
            stack[#stack] = nil
        end
    end
    return #stack == 0
end

local function problem27()
    print("Problem 27:")
    assert(isBalanced("([])[]({})")  == true,  "assertion failed")
    assert(isBalanced("([)]")        == false, "assertion failed")
    assert(isBalanced("(((()")       == false, "assertion failed")
    assert(isBalanced("")            == true,  "assertion failed")
    assert(isBalanced("{[]}")        == true,  "assertion failed")
    print("All Tests Passed")
end

return { problem27 = problem27 }

File: c/problem_27.c
#include "problem_27.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

int isBalanced(const char *s) {
    char stack[1024];
    int top = 0;
    for (; *s; s++) {
        char c = *s;
        if (c == '(' || c == '[' || c == '{') {
            stack[top++] = c;
        } else if (c == ')' || c == ']' || c == '}') {
            if (top == 0) return 0;
            char open = stack[--top];
            if ((c == ')' && open != '(') ||
                (c == ']' && open != '[') ||
                (c == '}' && open != '{')) return 0;
        }
    }
    return top == 0;
}

void problem27(void) {
    printf("Problem 27:\n");
    assert( isBalanced("([])[]({})" ));
    assert(!isBalanced("([)]")       );
    assert(!isBalanced("(((()")      );
    assert( isBalanced("")           );
    assert( isBalanced("{[]}")       );
    printf("All Tests Passed\n");
}
